core/src/main/scala/org/bykn/bosatsu/TotalityCheck.scala

```diff
@@ -454,24 +617,49 @@ case class TotalityCheck(inEnv: TypeEnv[Any]) {
             u.exists(isTopCheap(_))
         }
 
-      def subset(a0: Pattern[Cons, Type], b0: Pattern[Cons, Type]): Boolean =
-        isTopCheap(b0) || {
-
-          def loop(a: Pattern[Cons, Type], b: Pattern[Cons, Type]): Boolean =
-            (a, b) match {
-              case _ if a == b => true
-              case (_, Union(h, t)) => (h :: t).exists(loop(a, _))
-              case (Literal(Lit.Str(s)), sp@Pattern.StrPat(_)) => sp.matches(s)
-              case (s1@Pattern.StrPat(_), s2@Pattern.StrPat(_)) =>
-                strPatternSetOps.subset(s1, s2)
-              case (l1@Pattern.ListPat(_), l2@Pattern.ListPat(_)) =>
-                listPatternSetOps.subset(l1, l2)
-              case (Pattern.PositionalStruct(ln, lp), Pattern.PositionalStruct(rn, rp)) =>
-                (ln == rn) &&
-                  (lp.size == rp.size) &&
-                  (lp.zip(rp).forall { case (l, r) => loop(l, r) })
-              case _ => false
-            }
+      override def subset(a0: Pattern[Cons, Type], b0: Pattern[Cons, Type]): Boolean =
+        isTopCheap(b0) || relate(a0, b0).isSubtype
+
+      def relate(a0: Pattern[Cons,Type], b0: Pattern[Cons,Type]): Rel = {
+        def loop(a: Pattern[Cons, Type], b: Pattern[Cons, Type]): Rel =
+          (a, b) match {
+            case _ if a == b => Rel.Same
+            case (_, Union(h, t)) =>
+              urm.unionRelCompare(
+                Some(a),
+                Some(h),
+                Some(Pattern.union(t.head, t.tail)))
+            case (Union(_, _), _) => loop(b, a).invert
+            case (Literal(Lit.Str(s)), sp@Pattern.StrPat(_)) =>
+              // due to normalization, StrPat is not a literal
+              if (!sp.matches(s)) Rel.Disjoint
+              else Rel.Sub
+            case (Pattern.StrPat(_), Literal(Lit.Str(_))) =>
+              // due to normalization, StrPat is not a literal
+              loop(b, a).invert
+            case (s1@Pattern.StrPat(_), s2@Pattern.StrPat(_)) =>
+              strPatternSetOps.relate(s1, s2)
+            case (l1@Pattern.ListPat(_), l2@Pattern.ListPat(_)) =>
+              listPatternSetOps.relate(l1, l2)
+            case (Pattern.PositionalStruct(ln, lp), Pattern.PositionalStruct(rn, rp)) =>
+              if ((ln == rn) && (lp.size == rp.size)) {
+                (lp.zip(rp).foldLeft(Rel.Same: Rel) { case (acc, (l, r)) =>
+                  acc.lazyCombine(loop(l, r))
+                })
+              }
+              else Rel.Disjoint
+            case _ =>
+              val ta = isTop(a)
```

Comment:
Can we avoid using isTop here? If so, isTop could become as simple as relate.

Since the pattern is normalized, isTop may be the same as Wildcard here.